File: basemodule/src/main/java/com/victory/basemodule/download/DownLoaderHelper.java
package com.victory.basemodule.download;

import android.os.Environment;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

/**
 * @author Victory
 * @date 2019/6/19
 * DownLoaderHelper :
 */
public class DownLoaderHelper {

    private OkHttpClient okHttpClient;
    private static DownLoaderHelper mDownLoaderHelper;
    private Request request;


    public static DownLoaderHelper getDownLoaderHelper() {
        if (mDownLoaderHelper == null) {
            mDownLoaderHelper = new DownLoaderHelper();
        }
        return mDownLoaderHelper;
    }

    private OkHttpClient getOkHttpClient() {
        okHttpClient = new OkHttpClient.Builder().build();
        return okHttpClient;
    }

    private Request getRequest(String url) {
        request = new Request.Builder().url(url).build();
        return request;
    }

    /**
     * 默认下载事件
     * @param downloadurl
     * @param downLoadStatesInterface
     */
    public void downloaderFile(String downloadurl, final DownLoadStatesInterface downLoadStatesInterface) {
        getOkHttpClient().newCall(getRequest(downloadurl)).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                if (downLoadStatesInterface != null) {
                    downLoadStatesInterface.error(e.toString());
                }
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                InputStream is = null;
                byte[] buf = new byte[2048];
                int len = 0;
                FileOutputStream fos = null;
                // 储存下载文件的目录
                File dir = new File(Environment.getExternalStorageDirectory() + "");
                if (!dir.exists()) {
                    dir.mkdirs();
                }
                File file = new File(dir, "12345.apk");
                try {
                    is = response.body().byteStream();
                    long total = response.body().contentLength();
                    fos = new FileOutputStream(file);
                    long sum = 0;
                    while ((len = is.read(buf)) != -1) {
                        fos.write(buf, 0, len);
                        sum += len;
                        int progress = (int) (sum * 1.0f / total * 100);
                        // 下载中更新进度条
                        downLoadStatesInterface.loading(progress);
                    }
                    fos.flush();
                    // 下载完成
                    downLoadStatesInterface.success(true);
                } catch (Exception e) {
                    downLoadStatesInterface.error(e.toString());
                } finally {
                    try {
                        if (is != null) {
                            is.close();
                        }
                    } catch (IOException e) {

                    }
                    try {
                        if (fos != null) {
                            fos.close();
                        }
                    } catch (IOException e) {
                    }
                }
            }
        });
    }

    /**
     * 自定义下载事件
     * @param downloadurl
     * @param filepath
     * @param filename
     * @param downLoadStatesInterface
     */
    public void downloaderFile(String downloadurl, final String filepath, final String filename, final DownLoadStatesInterface downLoadStatesInterface) {
        getOkHttpClient().newCall(getRequest(downloadurl)).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                if (downLoadStatesInterface != null) {
                    downLoadStatesInterface.error(e.toString());
                }
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                InputStream is = null;
                byte[] buf = new byte[2048];
                int len = 0;
                FileOutputStream fos = null;
                // 储存下载文件的目录
                File dir = new File(filepath);
                if (!dir.exists()) {
                    dir.mkdirs();
                }
                File file = new File(dir, filename);
                try {
                    is = response.body().byteStream();
                    long total = response.body().contentLength();
                    fos = new FileOutputStream(file);
                    long sum = 0;
                    while ((len = is.read(buf)) != -1) {
                        fos.write(buf, 0, len);
                        sum += len;
                        int progress = (int) (sum * 1.0f / total * 100);
                        // 下载中更新进度条
                        downLoadStatesInterface.loading(progress);
                    }
                    fos.flush();
                    // 下载完成
                    downLoadStatesInterface.success(true);
                } catch (Exception e) {
                    downLoadStatesInterface.error(e.toString());
                } finally {
                    try {
                        if (is != null) {
                            is.close();
                        }
                    } catch (IOException e) {

                    }
                    try {
                        if (fos != null) {
                            fos.close();
                        }
                    } catch (IOException e) {
                    }
                }
            }
        });
    }

}

File: basemodule/src/main/java/com/victory/basemodule/download/DownLoaderHelper.kt
package com.victory.basemodule.download

import android.Manifest
import android.os.Environment
import android.util.Log
import com.victory.basemodule.constant.BaseConstant
import com.victory.basemodule.tools.PermissionUtil

import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream

import okhttp3.Call
import okhttp3.Callback
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response

/**
 * @author Victory
 * @date 2019/6/19
 * DownLoaderHelper :
 */
class DownLoaderHelper {

    private var okHttpClient: OkHttpClient? = null
    private var request: Request? = null

    private fun getOkHttpClient(): OkHttpClient {
        okHttpClient = OkHttpClient.Builder().build()
        return okHttpClient!!
    }

    private fun getRequest(url: String): Request {
        request = Request.Builder().url(url).build()
        return request!!
    }

    /**
     * 默认下载事件
     *
     * @param downloadurl
     * @param downLoadStatesInterface
     */
    fun downloaderFile(downloadurl: String, downLoadStatesInterface: DownLoadStatesInterface?) {

        if (PermissionUtil.isHasPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)){
            Log.e(BaseConstant.CommonTAG,"true")
        }else{
            PermissionUtil.getPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
            Log.e(BaseConstant.CommonTAG,"false")
            return
        }

        getOkHttpClient().newCall(getRequest(downloadurl)).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                downLoadStatesInterface?.error(e.toString())
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                var inputStream: InputStream? = null
                val buf = ByteArray(2048)
                var len = 0
                var fos: FileOutputStream? = null
                // 储存下载文件的目录
                val dir = File(Environment.getExternalStorageDirectory().toString() + "")
                if (!dir.exists()) {
                    dir.mkdirs()
                }
                val file = File(dir, "12345.apk")
                try {
                    inputStream = response.body()!!.byteStream()
                    val total = response.body()!!.contentLength()
                    fos = FileOutputStream(file)
                    var sum: Long = 0
                    len = (inputStream!!.read(buf))
                    while (len != -1) {
                        fos.write(buf, 0, len)
                        sum += len.toLong()
                        val progress = (sum * 1.0f / total * 100).toInt()
                        // 下载中更新进度条
                        downLoadStatesInterface!!.loading(progress)
                    }
                    fos.flush()
                    // 下载完成
                    downLoadStatesInterface!!.success(true)
                } catch (e: Exception) {
                    downLoadStatesInterface!!.error(e.toString())
                } finally {
                    try {
                        inputStream?.close()
                    } catch (e: IOException) {

                    }

                    try {
                        fos?.close()
                    } catch (e: IOException) {
                    }

                }
            }
        })
    }

    /**
     * 自定义下载事件
     *
     * @param downloadurl
     * @param filepath
     * @param filename
     * @param downLoadStatesInterface
     */
    fun downloaderFile(downloadurl: String, filepath: String, filename: String, downLoadStatesInterface: DownLoadStatesInterface?) {

        getOkHttpClient().newCall(getRequest(downloadurl)).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                downLoadStatesInterface?.error(e.toString())
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                var inputStream: InputStream? = null
                val buf = ByteArray(2048)
                var len = 0
                var fos: FileOutputStream? = null
                // 储存下载文件的目录
                val dir = File(filepath)
                if (!dir.exists()) {
                    dir.mkdirs()
                }
                val file = File(dir, filename)
                try {
                    inputStream = response.body()!!.byteStream()
                    val total = response.body()!!.contentLength()
                    fos = FileOutputStream(file)
                    var sum: Long = 0
                    len = inputStream!!.read(buf)
                    while (len != -1) {
                        fos.write(buf, 0, len)
                        sum += len.toLong()
                        val progress = (sum * 1.0f / total * 100).toInt()
                        // 下载中更新进度条
                        downLoadStatesInterface!!.loading(progress)
                    }
                    fos.flush()
                    // 下载完成
                    downLoadStatesInterface!!.success(true)
                } catch (e: Exception) {
                    downLoadStatesInterface!!.error(e.toString())
                } finally {
                    try {
                        inputStream?.close()
                    } catch (e: IOException) {

                    }

                    try {
                        fos?.close()
                    } catch (e: IOException) {
                    }

                }
            }
        })
    }

    companion object {
        private var mDownLoaderHelper: DownLoaderHelper? = null

        val downLoaderHelper: DownLoaderHelper
            get() {
                if (mDownLoaderHelper == null) {
                    mDownLoaderHelper = DownLoaderHelper()
                }
                return mDownLoaderHelper!!
            }
    }

}
